File: go/cryptoutil/cipher.go
package cryptoutil

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"github.com/jc-lab/jscp/go/payloadpb"
)

// Encrypt encrypts the plaintext using AES-GCM
func Encrypt(algorithm payloadpb.CryptoAlgorithm, dest *payloadpb.EncryptedMessage, key, authData, plaintext []byte) error {
	block, err := aes.NewCipher(key)
	if err != nil {
		return err
	}

	nonce := dest.Nonce
	if len(nonce) <= 0 {
		nonce = make([]byte, 12)
		if _, err := rand.Read(nonce); err != nil {
			return err
		}
		dest.Nonce = nonce
	}

	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return err
	}

	ciphertext := aesgcm.Seal(nil, nonce, plaintext, authData)
	dest.Ciphertext = ciphertext
	return nil
}

// Decrypt decrypts the ciphertext using AES-GCM
func Decrypt(algorithm payloadpb.CryptoAlgorithm, key []byte, authData []byte, input *payloadpb.EncryptedMessage) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	aesgcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	plaintext, err := aesgcm.Open(nil, input.Nonce, input.Ciphertext, authData)
	if err != nil {
		return nil, err
	}

	return plaintext, nil
}

File: node/src/crypto/cipher.ts
import * as proto from '../proto';
import { AES } from '@stablelib/aes';
import { GCM } from '@stablelib/gcm';

export interface Cipher {
    readonly type: proto.CipherAlgorithm;
    seal: (key: Uint8Array, nonce: Uint8Array, plaintext: Uint8Array, ad: Uint8Array) => Uint8Array;
    open: (key: Uint8Array, nonce: Uint8Array, ciphertext: Uint8Array, ad: Uint8Array) => Uint8Array;
}

export class AesGcmCipher implements Cipher {
    public readonly type: proto.CipherAlgorithm = proto.CipherAlgorithm.CipherAesGcm;

    seal(key: Uint8Array, nonce: Uint8Array, plaintext: Uint8Array, ad: Uint8Array): Uint8Array {
        const aead = new GCM(new AES(key, true));
        return aead.seal(nonce, plaintext, ad)!;
    }

    open(key: Uint8Array, nonce: Uint8Array, ciphertext: Uint8Array, ad: Uint8Array): Uint8Array {
        const aead = new GCM(new AES(key, false));
        const output = aead.open(nonce, ciphertext, ad);
        if (!output) {
            throw new Error('decrypt failed');
        }
        return output;
    }
}

export function getCipherAlgorithm(type: proto.CipherAlgorithm): Cipher {
    switch (type) {
        case proto.CipherAlgorithm.CipherAesGcm:
            return new AesGcmCipher();

        default:
            throw new Error(`unknown cipher algorithm: ${type}`);
    }
}